poweroff: nordic: streamline RAM retention strategy

nika-nordic opened the pull request:
retained_mem should be used for RAM retention management on nRF devices.
`sys_poweroff` should disable RAM retention for unused RAM blocks to achieve lowest power consumption in System OFF state.
Align `system_off` sample accordingly.

nika-nordic commented:
Adding `DNM` for now until downstream tests finalize

anangl commented on soc/nordic/common/Kconfig:
```diff
@@ -4,6 +4,9 @@
 	bool
 
+config HAS_NORDIC_RAM_CTRL
+	bool "nRF RAM control helper support"
```
Why is this prompt added? If I understand it correctly, this symbol is supposed to be selected by SoCs, not configured by users.

nika-nordic replied:
removed prompt

anangl commented on soc/nordic/common/poweroff.c:
```diff
@@ -17,8 +18,45 @@
+
+	/* Disable retention for all memory blocks */
+	nrfx_ram_ctrl_retention_enable_set(ram_start, ram_size, false);
+#endif
```
```suggestion
#endif /* defined(CONFIG_HAS_NORDIC_RAM_CTRL) */
```

nika-nordic committed:
soc: nordic: add common symbol for RAM control helper presence

Common symbol is defined by SoCs that support RAM control helper.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic committed:
drivers: retained_mem: nrf: make ram_ctrl variant explicit

Updated retained_mem driver variant exposes vendor-specific API,
rather than executing implicitly during boot.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic committed:
soc: nordic: poweroff: add retention management

RAM retention is disabled for all RAM blocks
before entering System OFF to achieve lowest power consumption.
RAM retention can still be enabled using retained_mem driver.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic committed:
samples: nordic: system_off: use retained_mem on all targets

Using bare-metal approach for retention configuration
is no longer compatible with sys_poweroff() implementation.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic committed:
soc: nordic: poweroff: disable remaining RAM on emul l05/l10 targets

nRF54L05 and nRF54L10 devices that are emulated on nRF54L15
needs to have manually disabled remaining RAM blocks
that are outside of their RAM region definitions.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic commented on samples/boards/nordic/system_off/CMakeLists.txt:
```diff
@@ -5,6 +5,6 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
 
 target_sources(app PRIVATE src/main.c)
-if (CONFIG_APP_USE_NRF_RETENTION OR CONFIG_APP_USE_RETAINED_MEM)
+if (CONFIG_APP_USE_RETAINED_MEM)
```
fixed

nordicjm commented on samples/boards/nordic/system_off/README.rst:
```diff
@@ -13,10 +13,10 @@ deep sleep on Nordic platforms.
-``CONFIG_APP_USE_NRF_RETENTION=y`` or ``CONFIG_APP_USE_RETAINED_MEM=y``
-state related to number of boots, number of times system off was entered,
-and total uptime since initial power-on are retained in a checksummed data structure.
+This sample can also demonstrate RAM retention.
+By selecting ``CONFIG_APP_USE_RETAINED_MEM=y`` state related to number of boots,
+number of times system off was entered, and total uptime since initial power-on
+are retained in a checksummed data structure.
```
All this code here https://github.com/zephyrproject-rtos/zephyr/pull/82262/files#diff-fca85c989d1e5dd6c706fe7388d14f46081d9de52b0603f0590cb458c3fa6342R38
Is all automatically handled simply by setting up a retained memory area then partitioning it (i.e. 100% of the area) with a retention part and enable checksum, you can see a setup and description here: https://docs.zephyrproject.org/latest/services/retention/index.html#devicetree-setup

nika-nordic replied:
ah, so we would need to start using `retention` subsystem rather than `retained_mem` driver to achieve this optimization.

Would it be possible to use `retained_mem` driver API for now in this sample, and switch to `retention` service in a separate, follow-up PR ?

nordicjm replied:
Yes, switching to retention is unrelated to this PR it's just in this PR where I noticed it's reimplementing what is already available

kartben commented:
@nika-nordic needs a rebase

nika-nordic commented:
> @nika-nordic needs a rebase

done
